pkg/server/server.go

```diff
@@ -831,6 +819,103 @@ func (s *ImmuServer) CreateDatabaseV2(ctx context.Context, req *schema.CreateDat
 	}, nil
 }
 
+func (s *ImmuServer) OpenDatabase(ctx context.Context, req *schema.OpenDatabaseRequest) (*schema.OpenDatabaseResponse, error) {
+	if req == nil {
+		return nil, ErrIllegalArguments
+	}
+
+	if !s.Options.GetAuth() {
+		return nil, ErrAuthMustBeEnabled
+	}
+
+	if req.Database == s.Options.defaultDBName || req.Database == SystemDBName {
+		return nil, ErrReservedDatabase
+	}
+
+	_, user, err := s.getLoggedInUserdataFromCtx(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("could not get loggedin user data")
+	}
+
+	//if the requesting user has admin permission on this database
+	if (!user.IsSysAdmin) &&
+		(!user.HasPermission(req.Database, auth.PermissionAdmin)) {
+		return nil, fmt.Errorf("the database '%s' does not exist or you do not have admin permission on this database", req.Database)
+	}
+
+	db, err := s.dbList.GetByName(req.Database)
+	if err != nil {
+		return nil, err
+	}
+
+	if !db.IsClosed() {
+		return nil, fmt.Errorf("%w: while opening database '%s'", ErrDatabaseAlreadyOpen, req.Database)
+	}
+
+	dbOpts, err := s.loadDBOptions(req.Database, false)
+	if err == store.ErrKeyNotFound {
+		return nil, database.ErrDatabaseNotExists
+	}
+	if err != nil {
+		return nil, fmt.Errorf("%w: while loading database settings", err)
+	}
+
+	db, err = database.OpenDB(req.Database, s.databaseOptionsFrom(dbOpts), s.Logger)
```

Comment:
We can end up in a race condition here - multiple callers could try to simultaneously open a database here. Between the check for `IsClosed` and OpenDB there's no lock.